asteroid/models/base_models.py

```diff
@@ -305,10 +305,12 @@ def forward(self, wav):
         tf_rep = self.postprocess_encoded(tf_rep)
         tf_rep = self.enc_activation(tf_rep)
 
-        est_masks = self.masker(tf_rep)
```

Comment:
To me all of this looks like a special case that ought not to handle in this generic forward() method. We don’t have to make forward() fit to every possible use case, otherwise it will become so generic that the abstraction hasn’t any meaning anymore. 

How about simply moving this entire mask application code into a method `apply_mask(tf_rep, est_masks) -> masked_tf_rep`? Or we can simply not use the parent forward() method at all.

Reply:
I was thinking about `apply_mask` too. This is better. 

I think `preprocess_masker_input` makes sense. If the filterbank is the STFT and you want some phase features, you don't want to mask them. It's pretty common IMO. Alternatively, we could have a `forward_masker` or `apply_masker` function instead of applying the masker directly, and this could contain additional feature extraction in subclasses.

Reply:
Maybe it would have been better to not implement these hooks like
that in the first place and only split into 3 methods `forward_{endoder,masker,decoder}` from the beginning... not sure

Reply:
> Or we can simply not use the parent forward() method at all.

The incentive to stick with the `forward` was to not copy paste the block of code that shapes the input/output.

> split into 3 methods `forward_{endoder,masker,decoder}`

I agree. Having a `forward_{encoder, masker, decoder}` offers more flexibility to the inheriting classes. In addition to these 3 blocks, I would add an `apply_mask` method to keep the mask application step separate from the previous blocks. Along these lines, the forward of `BaseEncoderMaskerDecoder` would look like:

```python
def forward(self, wav):
   shape = jitable_shape(wav)
   wav = _unsqueeze_to_3d(wav)

   tf_rep = self.forward_encoder(wav)
   est_masks = self.forward_masker(tf_rep)
   masked_tf_rep = self.apply_mask(tf_rep, est_masks)
   decode = self.forward_decoder(masked_tf_rep)

   reconstructed = pad_x_to_y(decoded, wav)
   return _shape_reconstructed(reconstructed, shape)

def forward_encoder(self, wav):
    # Computes time-frequency representation of `wav`
    # default: applies encoder and encoder activation
    # ...
    return tf_rep

def forward_masker(self, tf_rep):
    # Estimate masks based on `tf_rep`.
    # default: applies masker
    # ...
    return masks

def apply_masks(self, tf_rep, masks):
    # Applies `masks` on `tf_rep`
    return masks * tf_rep.unsqueeze(1)

def forward_decoder(self, masked):
    # Decodes masked time-frequency representation
    # default: applies decoder
    return decoded
```

With this structure, the only thing that is imposed to any class that inherits from `BaseEncoderMaskerDecoder` is the following diagram (again, in plain text):

```
(wav) --> Encoder --(tf_rep)----------------------> Mask_application ---(masked)---> Decoder ---> (decoded)
                            |                               1
                            |--> Mask_estimation --(masks)--|
```

Pre/post-processing would then be written directly in the inheriting classes.

For backward compatibility with classes that currently implements the `postprocess_*` hooks, we could call these in the `forward_*` and completely remove the hooks for v0.4? We would need to have tests ensuring backward compatibility for each of these models though :stuck_out_tongue:

Reply:
Also not sure but this does look good, and maybe more flexible. I'll need a bit of time to think about it. 

Apart from `DCUNet` and `DCCRN`, no model uses that IIRC, and we have no pretrained models for them, plus `DCCRN` is going to change.

Reply:
Reminds me of this :) 
<img width="778" alt="convtasnet" src="https://user-images.githubusercontent.com/175722/97972777-5662a400-1dc5-11eb-8544-f09378844646.png">
(from the Conv TasNet paper)

Reply:
I don't know this, what is ConvTasNet?

Reply:
Hello ! I agree from what i have seen in `DCUNet` and `DCCRN` the changes are small and straightforward. Having `forward_{endoder,masker,decoder}` seems like the good way to go :)